dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeerCache.h

```diff
@@ -34,21 +34,40 @@ using UniversalPageStoragePtr = std::shared_ptr<UniversalPageStorage>;
 class EndToSegmentId
 {
 public:
-    [[nodiscard]] std::unique_lock<std::mutex> lock();
+    [[nodiscard]] std::unique_lock<std::shared_mutex> writeLock();
+    [[nodiscard]] std::shared_lock<std::shared_mutex> readLock();
 
-    bool isReady(std::unique_lock<std::mutex> & lock) const;
+    bool isReady(std::unique_lock<std::shared_mutex> & lock) const;
+    bool isReady(std::shared_lock<std::shared_mutex> & lock) const;
 
     // The caller must ensure `end_key_and_segment_id` is ordered.
     // Called in `Segment::readAllSegmentsMetaInfoInRange`.
     void build(
-        std::unique_lock<std::mutex> & lock,
+        std::unique_lock<std::shared_mutex> & lock,
         std::vector<std::pair<DM::RowKeyValue, UInt64>> && end_key_and_segment_ids);
 
     // Given a key, return the segment_id that may contain the key
-    UInt64 getSegmentIdContainingKey(std::unique_lock<std::mutex> & lock, const DM::RowKeyValue & key);
+    template <typename LOCK>
+    UInt64 getSegmentIdContainingKey(LOCK & lock, const DM::RowKeyValue & key)
```

Comment:
Why not use `GenericLock`?

Reply:
GenericLock is defined under DeltaMerge module, whereas getSegmentIdContainingKey is in KVStore module.

Reply:
lambda can also use template now.

Reply:
It is kind of hard to do so. Because the lock type in this lambda depends on captured value `is_cache_ready`

Reply:
You can add `bool is_cache_ready` as template args.

like

```cpp
 auto protected_logic = [&]<typename Lock, bool is_cache_ready>(Lock & lock) {
        if constexpr (is_cache_ready)
        {
            current_segment_id
                = end_to_segment_id_cache->getSegmentIdContainingKey(lock, target_range.getStart().toRowKeyValue());
        }
        ...
}
```

call like 

```cpp
    if (is_cache_ready)
    {
        auto lock = end_to_segment_id_cache->readLock();
        return protected_logic.template operator()<std::shared_lock<std::shared_mutex>, true>(lock);
    }
    else
    {
        auto lock = end_to_segment_id_cache->writeLock();
        return protected_logic.template operator()<std::unique_lock<std::shared_mutex>, false>(lock);
    }
```

Reply:
https://en.cppreference.com/w/cpp/language/lambda#Syntax

Reply:
https://vittorioromeo.info/index/blog/cpp20_lambdas_compiletime_for.html

Reply:
https://en.cppreference.com/w/cpp/types/is_same you can also use `std::is_same_v`.

Reply:
Yes, I have tried this, but it is kind of ugly

Reply:
Yes, it is an idea, because those two lines are placed at the front and the last of the code. If we move them out, then  the lambda could be clearer.
I keep those code here in the lambda, because it is more readable. What is done in this lambda is that we will try fetch built cache with a shared lock, or build a cache with a unique lock. So IMO the build stage and the fetch stage could be better put in this lambda to achieve better readability of what the piece of code would do.

Reply:
IMO, now the style is not good.

1. Introduce a complex type GenericLock
2. Before lambda, we have already known the value of `is_cache_ready`, but we still need to check it several times in the lambda. 

Instead of passing the lock, pass `constexpr is_cache_ready` and acquire the lock inside the lambda maybe better.

Reply:
```cpp
auto protected_logic = [&](bool is_cache_ready) {
        std::variant<std::shared_lock<std::shared_mutex>, std::unique_lock<std::shared_mutex>> lock;
        if (is_cache_ready)
        {
            lock = end_to_segment_id_cache->readLock();
            current_segment_id = end_to_segment_id_cache->getSegmentIdContainingKey(
                std::get<std::shared_lock<std::shared_mutex>>(lock),
                target_range.getStart().toRowKeyValue());
        }
        else
        {
            lock = end_to_segment_id_cache->writeLock();
        }
```

Reply:
OK, I can take you last advice